### PR TITLE
Move pandas backend to backends/ directory

### LINKED ISSUE #2423: Move PANDAS backend to the backends/ directory
**Move the `pandas/` backend to `backends/pandas`.**

The current Ibis directory structure is a bit confusing, since backends are mixed with core modules. For example, it's not obvious to know if `pandas/` is a backend, or core functionality. Many of the backends live in the root directory, and some are grouped, like in a directory `sql/`. This makes things even harder and difficult to maintain. Also, ibis uses attributes to access the backends (e.g. `ibis.spark.connect`), and `ibis.spark` in code seems to be the `ibis/spark/` module, but it's not, making things confusing, and even creating problems with pytest.

We're moving backends to a standardized structure, in the `backends/` directory, one backend at a time (each backend in a separate pull request). We started by the OmniSciDB backend, and its PR can be a reference for the changes that need to be made here: #2392

The exact tasks to perform are:
- Move the backend from its original location to `backends/` (if the backend is in a subdirectory, like `sql/`, ignore the subdirectory and just have the backend inside `backends/`, so `sql/sqlite/` -> `backends/sqlite`)
- Move the content of `backends/<your-backend>/api.py` to `backends/<your-backend>/__init__.py`. Backends are using `api.py` for historical reasons, but the public backend API should be in the standard `__init__.py` file of the module.
- Update `ibis/__init__.py` so the backend is imported from the new module (`from ibis.<your-backend> import api` -> `from ibis import <your-backend>`)
- Look for places in the code where your backend is being imported, and update the imports. For imports inside the same backend, please use relative imports (e.g. `import ibis.<your-backend>.whatever` -> `import .whatever`). See the reference PR for examples.
- Update the documentation (in `doc/source/backends/<your-backend>.rst`) so the API is generated from the right location.

After performing the changes, the CI should be green, except for the packaging build (which requires changes in the conda-forge feedstock). You can ignore that error, those changes will be performed later for all moved backends.

### PR DESCRIPTION
**Pull request by joy-rosie**
closes #2423 

Discarding previous pull request #2429 as could not figure out a test failing. Most likely a circular import

### CODE REVIEW

**Comment by joy-rosie** [comment]
I think the tests relating the changes I made are passing, please let me know if there are any specific ones to look at.

Couple things to note:
1) https://github.com/ibis-project/ibis/pull/2440/commits/df38ad80402bc9fd3d8c2e2f3f01bedcf330d68a, I cannot make this import relative as it errors, I think this is due to a circular import but am not 100% sure, let me know if I can try something else
2) In the becnhmarks check: https://github.com/ibis-project/ibis/pull/2440/checks?check_run_id=1197740161#step:9:52, for some reason it is giving the following error:
```
[  0.00%] ···· Traceback (most recent call last):
                 File "/usr/share/miniconda/lib/python3.7/site-packages/asv/benchmark.py", line 860, in disc_modules
                   module = import_module(module_name)
                 File "/home/runner/work/ibis/ibis/.asv/env/9536bcdade9ef9f967582445231c0bf1/lib/python3.7/importlib/__init__.py", line 127, in import_module
                   return _bootstrap._gcd_import(name[level:], package, level)
                 File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
                 File "<frozen importlib._bootstrap>", line 983, in _find_and_load
                 File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
                 File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
                 File "<frozen importlib._bootstrap_external>", line 728, in exec_module
                 File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
                 File "/home/runner/work/ibis/ibis/benchmarks/benchmarks.py", line 6, in <module>
                   from ibis.backends.pandas.udf import udf
               ModuleNotFoundError: No module named 'ibis.backends.pandas'
```
Again I am not sure but this is happening, I tried this locally and seems to work fine.

**Comment by joy-rosie** [comment]
I just realised that the benchmark check is failing from it is running from master instead of this pull request

**Comment by joy-rosie** [comment]
@datapythonista, just tagging you to the new PR

File: benchmarks/benchmarks.py [hunk 1]
```diff
@@ -156,7 +156,7 @@ def setup(self):
             }
         )
 
-        t = ibis.pandas.connect({'df': self.data}).table('df')
+        t = ibis.backends.pandas.connect({'df': self.data}).table('df')
```
Comment (datapythonista) [inline benchmarks/benchmarks.py:159]: This is because of something very tricky and counter-intuitive in Ibis, but not correct.

If the file imported `ibis.pandas` (instead of `ibis`), the original `ibis.pandas` in this line would be the `ibis.pandas` module. But Since `import ibis` is used, this is the `pandas` attribute in `ibis/__init__.py` (which is the module, but that's irrelevant). That's quite bad practice IMHO, but what it means is that you need to revert this change.

File: docs/source/backends/pandas.rst [hunk 2]
```diff
@@ -4,12 +4,12 @@
 Ibis's pandas backend is available in core Ibis:
 
 Create a client by supplying a dictionary of DataFrames using
-:func:`ibis.pandas.connect`. The keys become the table names:
+:func:`ibis.backends.pandas.connect`. The keys become the table names:
```
Comment (datapythonista) [inline docs/source/backends/pandas.rst:7]: Same as before

File: docs/source/backends/pandas.rst [hunk 3]
```diff
@@ -4,12 +4,12 @@
...
 
 .. code-block:: python
 
    import pandas as pd
-   con = ibis.pandas.connect(
+   con = ibis.backends.pandas.connect(
```
Comment (datapythonista) [inline docs/source/backends/pandas.rst:12]: same, please revert

File: ibis/backends/pandas/aggcontext.py [hunk 4]
```diff
@@ -312,7 +312,7 @@ def compute_window_spec_none(_, obj):
 
 @compute_window_spec.register(dt.Interval)
 def compute_window_spec_interval(_, expr):
-    value = ibis.pandas.execute(expr)
+    value = ibis.backends.pandas.execute(expr)
```
Comment (datapythonista) [inline ibis/backends/pandas/aggcontext.py:315]: Same as before, can you revert please?

File: ibis/tests/all/test_temporal.py [hunk 5]
```diff
@@ -357,7 +357,7 @@ def test_to_timestamp(backend, con, unit):
     factor = unit_factors[unit]
 
     ts = ibis.timestamp('2018-04-13 09:54:11.872832')
-    pandas_ts = ibis.pandas.execute(ts).floor(unit).value
+    pandas_ts = ibis.backends.pandas.execute(ts).floor(unit).value
```
Comment (datapythonista) [inline ibis/tests/all/test_temporal.py:360]: Can you revert this too.

File: ibis/tests/backends.py [hunk 6]
```diff
@@ -168,7 +168,7 @@ class Pandas(Backend, RoundHalfToEven):
 
     @staticmethod
     def connect(data_directory: Path) -> ibis.client.Client:
-        return ibis.pandas.connect(
+        return ibis.backends.pandas.connect(
```
Comment (datapythonista) [inline ibis/tests/backends.py:171]: Also this.

File: ibis/tests/expr/test_operations.py [hunk 7]
```diff
@@ -106,7 +106,7 @@ class SpecialTable(ops.DatabaseTable):
         def output_type(self):
             return MyTableExpr
 
-    con = ibis.pandas.connect({})
+    con = ibis.backends.pandas.connect({})
```
Comment (datapythonista) [inline ibis/tests/expr/test_operations.py:109]: And this.

**Review by datapythonista — COMMENTED** [review]
Great work @joy-rosie, really nice. There are a couple of couple cases that better to change, but other of that look perfect.

**Comment by joy-rosie** [comment]
Sure, I can do that. Although, I have a question related the import. 

In `ibis/__init__.py` when we do `from ibis.backends import pandas`, my understanding is that it allows users to use `ibis.pandas` but when I tried locally, I could only use `ibis.backends.pandas`.

Could you help me understand what `from ibis.backends import pandas` exactly does?

**Comment by joy-rosie** [comment]
Just read https://github.com/ibis-project/ibis/pull/2440#discussion_r498865548.

So then my understanding is that if you do `import ibis` then you have access to `ibis.pandas` otherwise you have to use `ibis.backends.pandas`?

And for the user and outside `ibis/backends/pandas` we would prefer to use `ibis.pandas` unless it has explicitly imported `ibis.pandas...`?

**Comment by joy-rosie** [comment]
My last question is that why did that not error? If I used `ibis.backends.pandas` from only having run `import ibis` then should it not have errored? Because backends is never imported in `ibis/__init__.py`.

File: ibis/backends/pandas/execution/__init__.py [hunk 8]
```diff
@@ -0,0 +1,11 @@
+from .arrays import *  # noqa: F401,F403
```
Comment (datapythonista) [inline ibis/backends/pandas/execution/__init__.py:1]: Are all these needed for the move to work?

File: ibis/backends/pandas/dispatch.py [hunk 9]
```diff
@@ -4,11 +4,11 @@
 
 from multipledispatch import Dispatcher
 
-import ibis
+import ibis.client
```
Comment (datapythonista) [inline ibis/backends/pandas/dispatch.py:7]: Why is this change needed?

**Review by datapythonista — COMMENTED** [review]
Looks good, great job @joy-rosie.

Just couple of questions to understand better the changes, but it looks good.

There are couple of extra things that need to be done. We added `isort` to the validation now. Can you run `isort .` before merging master, and then again after merging master (the one before should help git fix the conflicts automatically for you).

Also, we added a `ci/recipe/meta.yaml` file, can you update the `ibis.pandas` path to `ibis.backends.pandas` (you can look for `ibis.backends.omniscidb` for reference).

Thanks!

File: ibis/backends/pandas/execution/__init__.py (hunk 8 above)
Comment (joy-rosie) [reply ibis/backends/pandas/execution/__init__.py:1]: not necessarily but they were there before so I thought it was worth keeping, I can test if the tests run without them if you like

File: ibis/backends/pandas/dispatch.py (hunk 9 above)
Comment (joy-rosie) [reply ibis/backends/pandas/dispatch.py:7]: this was not needed but thought it would be cleaner this way and a few other files do it this way as well, I can change this back if you like

File: ibis/backends/pandas/execution/__init__.py (hunk 8 above)
Comment (datapythonista) [reply ibis/backends/pandas/execution/__init__.py:1]: I didn't see them being moved in the diff, I thought you added them.

It's all right, just wanted to understand both points. Can you merge master and fix conflicts please?

**Comment by joy-rosie** [comment]
> Looks good, great job @joy-rosie.
> 
> Just couple of questions to understand better the changes, but it looks good.
> 
> There are couple of extra things that need to be done. We added `isort` to the validation now. Can you run `isort .` before merging master, and then again after merging master (the one before should help git fix the conflicts automatically for you).
> 
> Also, we added a `ci/recipe/meta.yaml` file, can you update the `ibis.pandas` path to `ibis.backends.pandas` (you can look for `ibis.backends.omniscidb` for reference).
> 
> Thanks!

To clarify, I need to merge from ibis/master into my branch first then run isort?

**Comment by datapythonista** [comment]
> To clarify, I need to merge from ibis/master into my branch first then run isort?

I'd do an isort first, I think that would avoid having to fix the conflicts manually. Then you can merge master, and isort again, to make sure the CI is green.

File: ibis/backends/pandas/execution/__init__.py (hunk 8 above)
Comment (joy-rosie) [reply ibis/backends/pandas/execution/__init__.py:1]: I merged master, fixed conflicts and pushed but it shows there are still conflicts and the CI is also not running

Comment (joy-rosie) [reply ibis/backends/pandas/execution/__init__.py:1]: sorry, give me a bit of time, I thought I did that but it doesnt seemed to have worked

**Comment by joy-rosie** [comment]
I think I created a bit of a mess but should be all good now

File: ibis/backends/pandas/execution/__init__.py [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+from .decimal import *  # noqa: F401,F403
+from .generic import *  # noqa: F401,F403
+from .join import *  # noqa: F401,F403
```
Comment (jreback) [inline ibis/backends/pandas/execution/__init__.py:4]: instead of adding the noqa

add

```
__all__ = ('decimal', 'join', .....)
```

**Review by jreback — CHANGES_REQUESTED** [review]

File: ibis/backends/pandas/execution/__init__.py (hunk 10 above)
Comment (joy-rosie) [reply ibis/backends/pandas/execution/__init__.py:4]: I tried to do that but got test failures, what exactly do you mean?

I replaced all the lines in that file by __all__ = (...) but that gives failures in the tests

Comment (jreback) [reply ibis/backends/pandas/execution/__init__.py:4]: like this: https://github.com/pandas-dev/pandas/blob/master/pandas/core/arrays/__init__.py

Comment (joy-rosie) [reply ibis/backends/pandas/execution/__init__.py:4]: I think I did what you have asked for, let me know if that works.

Out of curiosity, why do you want to do that? Is it for speed?

Comment (joy-rosie) [reply ibis/backends/pandas/execution/__init__.py:4]: Sorry I think I tried doing what you suggested then I get the following errors https://github.com/ibis-project/ibis/pull/2440#issuecomment-705865599

Comment (jreback) [reply ibis/backends/pandas/execution/__init__.py:4]: i c, this is not just a list of functions, but much more, we should not do this. 

instead we should figure out what is actually in the public api. ok remove the ``__all__`` and pls create an isseu to followup on this.

Comment (joy-rosie) [reply ibis/backends/pandas/execution/__init__.py:4]: I have made a new issue https://github.com/ibis-project/ibis/issues/2460 and removed `__all__` for now.

File: ibis/backends/pandas/dispatch.py (hunk 9 above)
Comment (joy-rosie) [reply ibis/backends/pandas/dispatch.py:7]: Can I resolve this conversation?

**Comment by jreback** [comment]
@joy-rosie sorry can you merge master again; we just had a pandas clean up PR go thru. this looks good. ping on green.

**Comment by joy-rosie** [comment]
> @joy-rosie sorry can you merge master again; we just had a pandas clean up PR go thru. this looks good. ping on green.

Sure, let me do that.

**Review by datapythonista — APPROVED** [review]
Amazing job @joy-rosie, looks perfect.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @joy-rosie very nice!
